### ISSUE TITLE
CP-2552 Version 2 planning

### ISSUE THREAD

**Issue by evanweible-wf** [Workiva/dart_dev#171]
**This is still a work-in-progress. Feedback is welcome and appreciated!**

---
### Simplified Configuration

Currently, configuration requires having a `tool/dev.dart` file that imports a config object and a function to hand execution off to the dart_dev executable.

A simplified and more discoverable approach would be to use a YAML file at the root (**`dart_dev.yaml`**) of the package (an approach that many tools have been adopting recently).

Each task would have a top-level entry in the YAML file where configuration for that task will live. Additionally, "local" tasks can be registered here.
### Task Lifecycle Hooks

One of the more powerful features of dart_dev is the ability to register `before` and `after` hooks to run for each task. This allows consumers to, for example, spin up an integration server prior to running the `test` task.

As mentioned in the above section, the plan is to move away from the `tool/dev.dart` file for configuration and instead use a YAML file. But, in order to support the aforementioned hooks, we have to be able to run a `.dart` file from the project. Instead of `tool/dev.dart` playing that role, we will look for a `tool/dart_dev_hooks.dart` file. The API for registering hooks will be more robust and will allow consumers to receive useful information about the task to be run like the package configuration and the command line options.

**tool/dart_dev_hooks.dart**

``` dart
import 'package:dart_dev/hooks.dart' show Config, DartDev, Tasks;

void main(List<String> args) {
  // Register hooks for a built-in dart_dev task.
  Tasks.test.registerHooks(
      before: (Config config) async {
        if (config.test.tags.contains('integration')) {
          // Spin up an integration server.
        }
      },
      after: (Config config) async {
        if (config.test.tags.contains('integration')) {
          // Spin down the integration server.
        }
      });

  // Register hooks for any task. This could be used to hook into
  // local tasks or to perform a generic action on every task.
  Tasks.any.registerHooks(before: (String taskName, Config config) async { ... });

  return DartDev.start(args);
}
```
### Discoverability / Guidance

One of the advantages of dart_dev is that it provides a consistent interface for tooling across. Each package can configure these tasks as desired and they can register additional local tasks. To assist contributors, dart_dev should make it a priority to display helpful messaging at all times. To be more specific: dart_dev should tell the CLI user what tasks are available and how the package has configured each task.
### Lighter Footprint

As this is a tool built on top of other tools, there are quite a few dependencies (`analyzer`, `coverage`, `dart_style`, `test`, etc). There are two main disadvantages to this:
1. Consumers cannot pick and choose which tasks to leverage - they get the dependencies for all of the tasks no matter what.
2. The `dart_dev` package becomes a restricting agent to the version constraint solver. If we are not actively watching our dependencies and widening our dependency ranges, then we may inadvertently cause pain for downstream consumers.

Dart_dev leverages these packages via their executables, which means consumers _have_ to explicitly specify them as a dependency anyway. This means it's fine for dart_dev to specify these dependencies as dev_dependencies and rely on consumers to specify the dependencies they want to use. This also gives consumers the freedom to choose their own version ranges for these dependencies. Since we only use the executables, we're less likely to be affected by breaking changes.
## Task List

> Version 2.0.0 will be built piece-by-piece (following the order laid out in the list below) and the [stacked pull requests approach](https://graysonkoonce.com/stacked-pull-requests-keeping-github-diffs-small/) will be used to make code review easier.
- [ ] **Documentation**
  - [ ] Simplified README
    - [ ] Installation
    - [ ] Basic Usage
    - [ ] Basic Configuration
  - [ ] Task Guides
    - [ ] Package Configuration
    - [ ] CLI Help/Usage
    - [ ] Examples
  - [ ] Improved help/usage test
- [ ] **Support `dart_dev.yaml` configuration**
  - [ ] API for parsing YAML and enforcing a schema
  - [ ] Class for storing and passing around the parsed YAML
- [ ] **Lenient arg parsing**
  - [ ] Capture unused/unknown args instead of throwing
- [ ] **Task Lifecycle**
  - [ ] Base class that defines standard task lifecycle hooks (before & after - some tasks will have additional lifecycle hooks)
  - [ ] API for consumers to register handlers for these hooks
- [ ] **Improved Testability**
  - [ ] Clear separation between the dart_dev task and the underlying executable - should be able to obtain the output from the dart_dev part and make assertions on it instead of actually calling through to the underlying executable for every test.
  - [ ] More unit tests, fewer integration tests.
- [ ] **Prune Dependencies**
  - [ ] Move as many dependencies as possible to `dev_dependencies`
- [ ] **Improved Tasks**
  - [ ] [**`analyze`** - runs the dartanalyzer (via the Dart SDK)](https://github.com/Workiva/dart_dev/issues/172)
    - [ ] Move `analyzer` from dependency to dev_dependency
    - [ ] `dart_dev.yaml` configuration
      - [ ] Libraries to analyze
    - [ ] Pass unused command line options to the `dartanalyzer` executable
    - [ ] Improved help/usage output
    - [ ] Detect un-analyzed files
  - [ ] [**`apply-license`** - copies LICENSE file contents to files as a header comment](https://github.com/Workiva/dart_dev/issues/173)
    - [ ] `dart_dev.yaml` configuration
      - [ ] License file path
      - [ ] Include/exclude globs
    - [ ] Increased file type support
    - [ ] Support updating existing license headers (`--update-from`)
  - [ ] [**`coverage`** - runs tests, collects coverage (via `coverage` package), and generates a report](https://github.com/Workiva/dart_dev/issues/174)
  - [ ] [**`docs`** - generates documentation (via `dartdoc` package)](https://github.com/Workiva/dart_dev/issues/175)
    - [ ] Move `dartdoc` from dependency to dev dependency
    - [ ] `dart_dev.yaml` configuration
      - [ ] TBD
    - [ ] Pass unused command line options to the `dartdoc` executable
  - [ ] [**`format`** - formats dart files (via `dart_style` package)](https://github.com/Workiva/dart_dev/issues/176)
    - [ ] Move `dart_style` from dependency to dev dependency
    - [ ] `dart_dev.yaml` configuration
      - [ ] Include/exclude globs
      - [ ] Line length
    - [ ] Pass unused command line options to the `dartfmt` executable
    - [ ] Improved help/usage output
  - [ ] [**`gen-test-runner`** - generates test runner file(s)](https://github.com/Workiva/dart_dev/issues/177)
  - [ ] [**`sort-imports`**](https://github.com/Workiva/dart_dev/issues/180)
    - [ ] Utilize `analysis_options.yaml` and `dart_dev.yaml` configuration to determine the set of files to operate on.
    - [ ] Sort into the following groups (each one alphabetized)
      - `dart:` imports
      - `package:` imports
      - `./` relative imports
    - [ ] `--check` for CI usage
  - [ ] [**`serve`** - runs `pub serve`](https://github.com/Workiva/dart_dev/issues/178)
    - [ ] `dart_dev.yaml` configuration
      - [ ] Map of directories to ports
      - [ ] Reverse proxying
    - [ ] Pass unused command line options to the `pub serve` executable
    - [ ] Improved help/usage output
  - [ ] [**`test`** - runs tests (via `test` package)](https://github.com/Workiva/dart_dev/issues/179)

---

> ### Misc (notes to self)
> - Support multiple `--pub-serve` options to allow easily serving multiple directories. There should also be a way to specify the port for each directory.
> - Support `pub-serve` in package configuration for every task (slightly different for the `serve` task), where it would be either a list of directories or a map of directories to ports.
> - Support already running `pub serve` processes to speed up tasks in a watch cycle (check to see if port is already bound, if so, assume the `pub serve` process is already running there, if not, start one)
> - Account for test suites that have all tests skipped (the message is "All tests skipped" instead of "All tests passed"). More generally, account for all variations of success/skipped/failed messages from the test package. Or find a better approach? Exit code? Contribute to test package?

**Comment by rmconsole-wf** [comment]
It looks like you have not included any tickets in the title of this issue. Please create a ticket for this issue, or click <a href='https://w-rmconsole.appspot.com/api/v1/issues/Workiva/dart_dev/171/addTicketToTitle'>here</a> to have Rosie create one for you.

**Comment by maxwellpeterson-wf** [comment]
> Additionally, "local" tasks can be registered here.

Would they no longer be auto discovered like they are now?

**Comment by georgelesica-wf** [comment]
> A simplified and more discoverable approach would be to use a YAML file...

Probably simpler, but I'm not sure it's more discoverable since the Dart-based configuration gives you completion. I can never remember what the various options are called so I just bring up the list and choose the one I want.

**Comment by georgelesica-wf** [comment]
> Instead of tool/dev.dart playing that role, we will look for a tool/dart_dev_hooks.dart file.

Is there a way to dynamically load / include files? It would be cool if these didn't need to all live in one file and could be auto-discovered like local tasks. Another thought, maybe the hooks should just be arbitrary programs and we pass data into them using the command line (maybe temporary JSON file, or even just pipe our stdout to their stdin and barf the data out that way. Then it would be possible to write hooks in other languages (like Bash or Python, in particular).

**Comment by evanweible-wf** [comment]
@georgelesica-wf Yeah I agree, this is the main thing I'm concerned about with the yaml config. I plan to add a validator to dart_dev that would be the first thing run for any ddev command, and my hope is that we can make the output helpful enough to offset the lack of autocompletion that the `dev.dart` approach gives you.

I've also toyed with the idea of a `config` task similar to `git`'s that could be used to update the `dart_dev.yaml` file as well as provide help output. For example..

```bash
$ ddev config
Tasks available for configuration:

    analyze
    coverage
    format
    test

$ ddev config analyze
Current configuration:
    analyze:
      entry-points:
        - lib/package.dart

Additional config options:
    analyze:
      exclude: [glob1, glob2, ...]

$ ddev config analyze.exclude
Current value: [none]

The `exclude` option is a list of globs that should be excluded. Any files matching at least one of the globs specified here will not be analyzed.

Example:
    analyze:
      exclude:
        - lib/generated/**
```

I'm not sure that interactive tool would be worth the effort, though. I'm hoping that a validator with enough helpful output could suffice. If you have any ideas for how to better address this concern, I'm all ears.. I'd really like to use the yaml config since it's a pattern that other dart-lang tools have adopted.

**Comment by georgelesica-wf** [comment]
Just a random idea. What if you could package a Dart_dev task independently? It would then be possible to depend on just the tasks you want, and to choose which versions you want. We could have some kind of "base" set, but it could be very small and have minimal dependencies. Others could be separate projects. I guess tasks could then depend on each other, not sure if that's a good or bad thing though.

**Comment by georgelesica-wf** [comment]
@evanweible-wf I like the `config` task idea in general. One problem is that Dart's YAML support is kind of weak. The built-in package only supports reading, at least as of a few months ago. There are a couple OSS libraries that support writing but I didn't have great confidence in them. When I tried to create the `link` task for ddev I used one of them and it didn't preserve order in the YAML file, so you'd get huge, meaningless diffs if you manually edited the file.

**Comment by evanweible-wf** [comment]
> Is there a way to dynamically load / include files? 

As far as I know, there isn't a good way to dynamically load or include files.

> It would be cool if these didn't need to all live in one file and could be auto-discovered like local tasks.

Yeah I agree, there's no reason we couldn't support that.

> Another thought, maybe the hooks should just be arbitrary programs and we pass data into them using the command line (maybe temporary JSON file, or even just pipe our stdout to their stdin and barf the data out that way. Then it would be possible to write hooks in other languages (like Bash or Python, in particular).

Being able to serialize the state of a dart_dev task as JSON so that it can be piped to any script instead of just Dart is an interesting idea, I'll noodle on that. I like the idea of being able to receive strongly-typed objects when writing these hooks, though. If interfacing with another language is necessary, the Dart hook could always start up that process on it's own.

> Just a random idea. What if you could package a Dart_dev task independently? It would then be possible to depend on just the tasks you want, and to choose which versions you want. We could have some kind of "base" set, but it could be very small and have minimal dependencies. Others could be separate projects. I guess tasks could then depend on each other, not sure if that's a good or bad thing though.

One of my goals for v2 is to remove the dependency on the underlying executables like `coverage`, `test`, etc. They would still be dev deps, just not normal deps. Consumers will be required to add those dependencies manually if they'd like to use that task. This sort of accomplishes the same thing by letting the consumer use the minimal set of deps. Also, it lets the consumer have control over which version of the underlying executable is installed instead of being restricted to a range defined in dart_dev.

Would that approach address your concern for minimal dependency sets?

(This approach does open up some other potential issues - in particular, not defining a range on an underlying executable. A consumer could pull in a newer version of the `coverage` package that isn't compatible with dart_dev's coverage task. But on the other hand, it's difficult to keep up with all of the releases of those packages and update the range in dart_dev, and the CLI for these packages are more susceptible to breaking changes since they're always considered part of the "public api" so we could hit that issue even with dart_dev defining a supported range.)

> I like the config task idea in general. One problem is that Dart's YAML support is kind of weak. The built-in package only supports reading, at least as of a few months ago. There are a couple OSS libraries that support writing but I didn't have great confidence in them. When I tried to create the link task for ddev I used one of them and it didn't preserve order in the YAML file, so you'd get huge, meaningless diffs if you manually edited the file.

Yeah, I remember that. I wouldn't consider the config task part of v2 MVP anyway, so maybe it's something that we just circle back on. We can still provide plenty of information by validating the config without the interactive part.

**Comment by georgelesica-wf** [comment]
> I like the idea of being able to receive strongly-typed objects when writing these hooks, though. If interfacing with another language is necessary, the Dart hook could always start up that process on it's own.

I agree that types should be preserved wherever reasonable. We might consider a task factory produces instances of a simple task that just starts up a script and sends it some data (lots of options for "how"). In some sense I'm thinking of a more general version of the "local" tasks we've got right now.

> Would that approach address your concern for minimal dependency sets?

I agree with you. I was more thinking about ways to extend dart_dev with additional tasks. One of the things I really like about [Mix](http://elixir-lang.org/getting-started/mix-otp/introduction-to-mix.html) is that libraries can automagically add tasks.

So, say you have Ecto (kind of an ORM) installed as a dependency in your project. You [automatically get](https://hexdocs.pm/ecto/getting-started.html) Mix tasks that are defined within the Ecto library in your project.

If we supported something like this it would also be possible to define (or maybe even override, that's another discussion) tasks through a pub dependency, meaning tasks themselves could be distributed / versioned through pub.

I guess I'm really proposing one idea that has two different use cases: provide a way to discover tasks that are exported by dependencies of a project that uses dart_dev.
